ethcore/private-tx/src/state_store.rs

```diff
@@ -0,0 +1,165 @@
+
+/// Type of the stored reques
+pub enum RequestType {
+	/// Verification of private transaction
+	Verification(Arc<VerifiedPrivateTransaction>),
+	/// Creation of the private transaction
+	Creation(SignedTransaction),
+}
+
+#[derive(Clone, PartialEq)]
+enum RequestState {
+	Syncing,
+	Ready,
+}
+
+struct StateRequest {
+	request_type: RequestType,
+	request_hashes: HashSet<H256>,
+	state: RequestState,
+}
+
+/// Wrapper over storage for the private states
+pub struct PrivateStateStorage {
+	private_state_db: Arc<PrivateStateDB>,
+	requests: RwLock<Vec<StateRequest>>,
+	syncing_hashes: RwLock<HashMap<H256, Instant>>,
+}
+
+impl PrivateStateStorage {
+	/// Constructs the object
+	pub fn new(db: Arc<KeyValueDB>) -> Self {
+		PrivateStateStorage {
+			private_state_db: Arc::new(PrivateStateDB::new(db)),
+			requests: RwLock::new(Vec::new()),
+			syncing_hashes: RwLock::default(),
+		}
+	}
+
+	/// Checks if ready for processing requests exist in queue
+	pub fn requests_ready(&self) -> bool {
+		let requests = self.requests.read();
+		requests.iter().find(|r| r.state == RequestState::Ready).is_some()
+	}
+
+	/// Signals that corresponding private state retrieved and added into the local db
+	pub fn state_sync_completed(&self, synced_state_hash: &H256) {
+		let mut syncing_hashes = self.syncing_hashes.write();
+		syncing_hashes.remove(synced_state_hash);
+		self.mark_hash_ready(synced_state_hash);
+	}
+
+	/// Returns underlying DB
+	pub fn private_state_db(&self) -> Arc<PrivateStateDB> {
+		self.private_state_db.clone()
+	}
+
+	/// Store a request for state's sync and later processing, returns new hashes, which sync is required
+	pub fn add_request(&self, request_type: RequestType, request_hashes: HashSet<H256>) -> Vec<H256> {
+		let request = StateRequest {
+			request_type: request_type,
+			request_hashes: request_hashes.clone(),
+			state: RequestState::Syncing,
+		};
+		let mut requests = self.requests.write();
+		requests.push(request);
+		let mut new_hashes = Vec::new();
+		for hash in request_hashes {
+			let mut hashes = self.syncing_hashes.write();
+			if hashes.insert(hash, Instant::now() + Duration::from_millis(MAX_REQUEST_SESSION_DURATION)).is_none() {
+				new_hashes.push(hash);
+			}
+		}
+		new_hashes
+	}
+
+	/// Drains ready requests to process
+	pub fn drain_ready_requests(&self) -> Vec<RequestType> {
+		let mut requests_queue = self.requests.write();
+		let mut drained = Vec::new();
+		let mut i = 0;
+		while i != requests_queue.len() {
+			if requests_queue[i].state == RequestState::Ready {
+				let request = requests_queue.remove(i);
+				drained.push(request.request_type);
+			} else {
+				i += 1;
+			}
+		}
+		drained
+	}
+
+	/// State retrieval timer's tick
+	pub fn tick(&self, logging: &Option<Logging>) {
+		let mut syncing_hashes = self.syncing_hashes.write();
+		let current_time = Instant::now();
+		syncing_hashes
+			.iter()
+			.filter(|&(_, expiration_time)| *expiration_time >= current_time)
```

Comment:
Not an issue, but probably some optimization for future PRs - there's a [`LinkedHashMap`](https://docs.rs/linked-hash-map/0.5.2/linked_hash_map/) struct - I've used it in similar cases (for detecting stale requests) to avoid iterating the whole `HashMap`.